ofproto/ofproto.c

```diff
@@ -5620,6 +5620,87 @@ handle_update_map(struct ofconn *ofconn, const struct ofp_header *oh)
     return error;
 }
 
+static enum ofperr
+handle_dump_map (struct ofconn *ofconn, const struct ofp_header *oh)
+{
+    struct ofproto *ofproto = ofconn_get_ofproto(ofconn);
+    enum ofperr error;
+    error = reject_slave_controller(ofconn);
+    if (error) {
+        return error;
+    }
+
+    struct ol_dump_map_request msg;
+    const ovs_be16 *map_ids;
+    error = ofputil_decode_dump_map_request(&msg, &map_ids, oh);
+
+    if (error) {
+        return error;
+    }
+
+    struct ubpf_vm *vm = ubpf_vms_lookup(ofproto, msg.filter_prog);
+    if (!vm) {
+        VLOG_WARN_RL(&rl,
+                     "The referenced filter program could not be found.");
+        return OFPERR_OFPBRC_EPERM;
+    }
+
+    if (msg.nb_maps <= 0) {
+        VLOG_WARN_RL(&rl,
+                     "Number of maps should be equal at least 1.");
+        return OFPERR_OFPBRC_EPERM;
+    }
+
+    struct ubpf_map **maps = malloc(msg.nb_maps * sizeof(struct ubpf_map *));
+    for(int i = 0; i < msg.nb_maps; i++) {
+        if (map_ids[i] >= vm->nb_maps) {
+            VLOG_WARN_RL(&rl,
+                         "The map with provided identifier does not exist.");
+            return OFPERR_OFPBRC_EPERM;
+        }
+
+        const char *name = vm->ext_map_names[map_ids[i]];
+        if (!name) {
+            VLOG_WARN_RL(&rl,
+                         "The map with the given ID does not exist.");
+            return OFPERR_OFPBRC_EPERM;
+        }
+
+        maps[i] = ubpf_lookup_registered_map(vm, name);
+        if (!maps[i]) {
+            VLOG_WARN_RL(&rl,
+                         "The referenced map of filter prog could not be found.");
+            free(maps);
+            return OFPERR_OFPBRC_EPERM;
+        }
+    }
+
+    void **data = malloc(msg.nb_maps * sizeof(void *));
+    unsigned int *nb_elems = malloc(msg.nb_maps * sizeof(unsigned int));
```

Comment:
You can remove all heap allocations in this function to use the stack instead.

Reply:
Done.

Reply:
You should also be able to remove the last `malloc` for `data[i]`.

Reply:
Hmm.., I don't know in advance how big should be the second dimension. That's why I use malloc.

Reply:
Some maps could have 1000 items other 1.

Reply:
Ah, right. That makes sense.